### ISSUE TITLE
counting pairs is inaccurate for repeating tokens?

### ISSUE THREAD

**Issue by NiceGuySaysHi** [karpathy/minbpe#51]
I just noticed that counting pairs might be slightly inaccurate for a lot of repeating tokens. For example in the sequence 1, 1, 1, 1 the pair (1, 1) gets counted 3 times, but we can only replace it two times. This is slightly inaccurate because for example in the sequence 1, 1, 1, 1, 2, 1, 2, 1, 2 we might choose the pair (1, 1) over (1, 2)

**Comment by NiceGuySaysHi** [comment]
Yes, i think it is not that bad in a real life scenario i guess. However look at what happens in our example: 1, 1, 1, 1, 2, 1, 2, 1, 2 -> 3,3,2,1,2,1,2. If we pick (1,2) now, then we can only substitute it 2 times. In the original sequence we could have substituted it 3 times. I'm not sure how bad it is, but i noticed (after writing this issue), that this exact problem is mentioned in the paper from the first issue. They also use the same counting function as in this repo and basically say that this is not that important. But i think you should have it in the Back of your mind

**Comment by zouharvi** [comment]
We mention it [in the paper page 3 top left](https://aclanthology.org/2023.findings-acl.38/) and have an algorithm with the adjustment on the last page (screenshot here). We were particular about it in the paper because it does have *formal* implications regarding the proofs. For them we want the quantity to be how much is compressed in the next step because pair frequency is not that.

Libraries usually fix it but imo this does not have a big impact *practically* on natural language (could be if there are super long repeated strings in the data such as `=============` due to inflated counts). The original paper by Rico Sennrich does not have this adjustment.

![Screenshot_20240302-204336~2](https://github.com/karpathy/minbpe/assets/7661193/deda88d5-0402-44ec-b344-b34b1d05c07d)

**Comment by NiceGuySaysHi** [comment]
yes, thank's for your reply. I already looked at your paper which you already linked in the first issue. Personally i think the algorithm is quite nice and i am currently implementing it inside c++ and call it in python using ctypes. I hope to get a much faster implementation that way. If you want to, i can link it to you if i am ready.

**Comment by Majdoddin** [comment]
I submitted PR #51 for it.
